tools/src/main/java/org/terracotta/utilities/io/Files.java

```diff
@@ -796,6 +801,101 @@ private static Path copyInternal(Path source, Path target, CopyOption[] options)
+   * or another exception being thrown.
+   * @param stream the {@code InputStream} to be copied to {@code target}
+   * @param target the path of the file into which the {@code InputStream} is written
+   * @param options {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} can be used to
+   *    delete {@code target} if it exists; other options are not supported
+   * @return the number of bytes copied
+   * @throws DirectoryNotEmptyException if {@code target} is a non-empty directory and
+   *          {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} is specified
+   * @throws FileAlreadyExistsException if {@code target} exists and
+   *          {@link StandardCopyOption#REPLACE_EXISTING REPLACE_EXISTING} is not specified
+   * @throws IOException if the copy operation fails
+   * @throws NullPointerException if {@code stream} or {@code target} is null
+   * @throws UnsupportedOperationException if {@code options} contains an unsupported value
+   * @throws SecurityException if permission to access the target file/directory is not sufficient
+   * @see java.nio.file.Files#copy(InputStream, Path, CopyOption...)
+   */
+  public static long copy(InputStream stream, Path target, CopyOption... options) throws IOException {
+    Objects.requireNonNull(stream, "stream must be non-null");
+    Objects.requireNonNull(target, "target must be non-null");
+
+    Set<CopyOption> copyOptions = new LinkedHashSet<>(Arrays.asList(options));
+    if (!EnumSet.of(REPLACE_EXISTING).containsAll(copyOptions)) {
+      copyOptions.remove(REPLACE_EXISTING);
+      throw new UnsupportedOperationException("Unsupported option(s) specified - " + copyOptions);
+    }
+    boolean replace = copyOptions.contains(REPLACE_EXISTING);
+
+    Path resolvedTarget = target.toAbsolutePath();
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug("copy(InputStream, {}, {})", resolvedTarget, Arrays.toString(options));
+    }
+
+    SecurityManager securityManager = System.getSecurityManager();
+    if (securityManager != null) {
+      securityManager.checkPermission(new FilePermission(resolvedTarget.toString(), "read,write" + (replace ? ",delete" : "")));
+    }
+
+    /*
+     * An attempt to open an OutputStream on a target that is a directory, empty or not, results in
+     * an AccessDeniedException (at least on Windows) so we need to delete a non-empty directory target
+     * _before_ attempting to open a stream on the target.  That leaves an unavoidable race window
+     * between the deletion and the stream open resulting in the possibility of fault on open.
+     *
+     * We also delete a file target being replaced to avoid AccessDeniedException caused by OS process
+     * interference during open.  This is subject to the same race window.
+     */
+    if (replace) {
+      // Throws a DirectoryNotEmptyException if a directory is not empty ...
+      deleteIfExists(resolvedTarget);
```

Comment:
An alternative to deleting both directories and files is to delete just directories and attempt to rely on `newOutputStream(..., CREATE, TRUNCATE_EXISTING)` to _replace_ existing files.  That would complicate acquisition of the `OutputStream` but _might_ work; if it did work, it would close the race window described in the comments for file replacement.  Symbolic links might complicate matters and would require some experimentation.